### PR TITLE
[IOTDB-2988] Cache leader in session

### PR DESCRIPTION
**Pull request by MrQuansy**

### CODE REVIEW

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/execution/QueryExecution.java [hunk 1]
```diff
@@ -360,6 +359,41 @@ private void initResultHandle() {
...
+          tsstatus.setCode(TSStatusCode.NEED_REDIRECTION.getStatusCode());
+          for (TEndPoint endPoint : redirectNodeList) {
+            subStatus.add(
+                StatusUtils.getStatus(TSStatusCode.NEED_REDIRECTION).setRedirectNode(endPoint));
+          }
+          tsstatus.setSubStatus(subStatus);
+        }
+      } else {
+        // single device
+        tsstatus.setRedirectNode(redirectNodeList.get(0));
```
Comment (OneSizeFitsQuorum) [inline server/src/main/java/org/apache/iotdb/db/mpp/plan/execution/QueryExecution.java:390]: Set `NEED_REDIRECTION` ?

**Review by OneSizeFitsQuorum — COMMENTED** [review]
Can we confirm that the first one should be the leader?

**Comment by xingtanzjr** [comment]
LGTM

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/execution/QueryExecution.java (hunk 1 above)
Comment (MrQuansy) [reply server/src/main/java/org/apache/iotdb/db/mpp/plan/execution/QueryExecution.java:390]: > 
We get data partition info from config nodes or local partition cache. Config nodes will set the leader at the first one in `TRegionReplicaSet`. When the leader has been transferred, we will update local partition cache afterwards.

Comment (MrQuansy) [reply server/src/main/java/org/apache/iotdb/db/mpp/plan/execution/QueryExecution.java:390]: Session can use TSStatus.isSetRedirectNode to check if the redirect TEndPoint is set to cache locally.

Comment (OneSizeFitsQuorum) [reply server/src/main/java/org/apache/iotdb/db/mpp/plan/execution/QueryExecution.java:390]: You could add some comments to these two questions both.

**Review by OneSizeFitsQuorum — APPROVED** [review]
LGTM

**Pull request merged**